### PR TITLE
ML morph implementation

### PR DESCRIPTION
**Pull request by leyangloh**
Complete pipeline for lizard toepad landmark prediction using dlib shape predictors     
  (ml-morph) with YOLO-detected bounding boxes. Compares two detection approaches:
                                                                                          
  - Baseline: Standard YOLO detect (axis-aligned bounding boxes, 6-class)                 
  - OBB: YOLO-OBB (oriented bounding boxes) with flip inference strategy

  Results: Baseline (H5) vs OBB (Axis-Aligned)
  - Toe: Baseline 83.27 px → OBB 30.72 px (2.7x better)
  - Finger: Baseline 40.64 px → OBB 38.25 px (1.06x better)

  Test error = average pixel deviation between predicted and ground-truth landmarks (144
  hyperparameter configs searched per condition).

  OBB wins both categories. Oriented bounding boxes converted to axis-aligned rects with
  30% padding produce tighter, more relevant crops for the shape predictor.

  Failed Experiment: Crop+Rotate OBB

  Attempted to crop and rotate images so OBB becomes upright with 10% padding. Results
  were significantly worse (toe: 92.02, finger: 76.11) due to rotation interpolation
  artifacts and landmark transformation issues. See docs/EXPERIMENT_CROP_ROTATE_OBB.md.

  What's Included

  - ml-morph framework: dlib shape predictor training with TPS→XML→YOLO bbox pipeline
  - Preprocessing scripts: tps_to_xml.py, generate_yolo_bbox_xml.py,
  create_noflip_obb_dataset.py
  - Inference: predict_landmarks_flip.py — end-to-end OBB detection + landmark prediction
  with flip strategy
  - Hyperparameter search: 144-config grid search over tree_depth, cascade_depth, nu,
  num_trees
  - SLURM sbatch files: For preprocessing, training, and hyperparameter search (baseline,
  OBB, OBB-aligned)
  - Documentation: COMPARISON_BASELINE_VS_OBB.md, EXPERIMENT_CROP_ROTATE_OBB.md

### CODE REVIEW

File: scripts/visualization/visualize_progression.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    
+    cv2.imwrite(str(output_path), vis_img)
+    print(f"Saved YOLO-OBB visualization to {output_path}")
+
+def main():
+    # Paths
+    img_path = Path("/storage/ice-shared/cs8903onl/miami_fall_24_jpgs/1001.jpg")
+    bbox_model = Path("/home/hice1/yloh30/scratch/Lizard_Toepads/yolo_bounding_box.pt")
+    obb_model = Path("/home/hice1/yloh30/scratch/Lizard_Toepads/runs/obb/H1_obb_2class2/weights/best.pt")
+    output_dir = Path("inference_results/progression")
```
[the remark this answers is not in the document]
Comment (JunlingZhuang) [reply scripts/visualization/visualize_progression.py:109]: Hi Leyang, I think what copliot said might be right, some of them is hardcoded and in your user path. We may wnat to have config driven script also there are a centralized download yolo model donwload script `scripts/download_models.py` You might just want to add YOLO OBB model download code there

File: configs/H5_obb_noflip.yaml
Comment (JunlingZhuang) [inline configs/H5_obb_noflip.yaml:1]: Woud you mind take a look existed H5 Config file. I feel like this missed a lot of vars as it might be hardcode in your modified script

**Comment by JunlingZhuang** [comment]
I actually not quite sure how to reduplicate the preproccess and traning script as there are lots of new script  without step by step doc. That makes the PR a little bit difficult for others to use. And may I know how those new scripts coupling/works with  existed scripts. It might be good for reference #8 , there are some update in README.md to point out how to use the new scripts

File: scripts/extract_id_from_yolo.py [hunk 2]
```diff
@@ -342,7 +342,7 @@ def get_opt(name: str, default):
         return default
 
     image_dir = get_opt('image-dir', None)
-    label_dir = get_opt('label-dir', None)
+    label_dir = get_opt('labels-dir', None)
     output_dir = get_opt('output-dir', 'data/processed')
```
[the remark this answers is not in the document]
Comment (JunlingZhuang) [reply scripts/extract_id_from_yolo.py:346]: Any reasons we want to break current script? It will casue conflcts with other collaborators for keeping using this script

File: docs/assets/bilateral_detection_plan/conf_Iou_treshhold_adjust/1004.jpg
Comment (JunlingZhuang) [inline docs/assets/bilateral_detection_plan/conf_Iou_treshhold_adjust/1004.jpg:1]: Recover these image files if possible, these are referenced in some progress update doc

**Comment by leyangloh** [comment]
> I actually not quite sure how to reduplicate the preproccess and traning script as there are lots of new script without step by step doc. That makes the PR a little bit difficult for others to use. And may I know how those new scripts coupling/works with existed scripts. It might be good for reference #8 , there are some update in README.md to point out how to use the new scripts

This should be addressed with [RUN_EXPERIMENTS_STEP_BY_STEP.md](https://github.com/Human-Augment-Analytics/Lizard_Toepads/blob/leyang/ml-morph/docs/RUN_EXPERIMENTS_STEP_BY_STEP.md)

**Review by JunlingZhuang — APPROVED** [review]
